app/waf/src/controllers/application.controller.ts

```diff
@@ -155,4 +173,82 @@ export class ApplicationController {
   async deleteById(@param.path.string('id') id: string): Promise<void> {
     await this.applicationRepository.deleteById(id);
   }
+
+  @post(prefix + '/applications/{id}/deploy', {
+    responses: {
+      '204': {
+        description: 'Application deploy success',
+      },
+    },
+  })
+  async deployById(@param.path.string('id') id: string): Promise<Object> {
+    let application = await this.applicationRepository.findById(id);
+
+    if (application.services.length === 0) {
+      throw new HttpErrors.UnprocessableEntity(
+        'No service in Application ' + application.id,
+      );
+    }
+
+    let params: {[key: string]: Object} = {
+      application: application,
+    };
+
+    let tenantAssocs = await this.tenantAssociationRepository.find({
```

Comment:
Since we know how to build the filter, the following seems more clear. 

```
    let tenantAssocs = await this.tenantAssociationRepository.find({
      limit: 1, 
      where: {
        tenantId: {
          inq: [application.tenantId, 'default']
        },
      },
    });

    if (tenantAssocs.length === 0) {
      throw new HttpErrors.UnprocessableEntity(
        'No ADC associated with Application ' + application.id,
      );
    }
```

But, 

> Lookup associated ADCs by tenantId first. If no ADC associated, then lookup the default ADC for all tenants instead. If no default ADC, that means we are not able to deploy configuration for this application.

I don't think there should be a ADC for all tenant. User should appoint one of them for deploying, otherwise, fails it. 
What's more, 

`params.adc = adcs[0];` may be a problem. When there're multiple ADCs, we should let user to select one of them instead of hardcoding to [0].

Reply:
The default ADC may be an use case. If we request every tenants to specify one ADC, that means we are attempt to import all tenants in to our db, but I do not think all of us are willing to do that. Anyway, it is not a programming question.

Since we are query ADC by its primary id, we can assume only one will be returned.

Reply:
OK for the current 1:1 model.

Reply:
Done